### ISSUE TITLE
A neo4j database error while creating nodes

### ISSUE THREAD

**Issue by zhangwenxuan322** [neo4j-contrib/neo4j-spark-connector#131]
My tools' version:
- Spark:2.3
- Neo4j:3.4.18
- neo4j-spark-connector:2.4.1-M1

I have a cluster with one master node and two slave nodes. Let's call these two slave nodes 'worker01' and 'worker02'. The total data amount was around 1 million lines in a spark data frame. 
While using `Neo4jDataFrame.createNodes()`, everything worked fine until 'worker01' node having the following error:
`Lost task 24.0 in stage 5.0 (TID 56, worker01, executor 5): org.neo4j.driver.v1.exceptions.DatabaseException: LockResult[NOT_LOCKED, Unable to acquire shared lock: Cannot access 'RequestContext[machineId=3, lastAppliedTransaction=236, eventIdentifier=10, hashCode=-1588610088, epoch=564542915457372, checksum=327561697546]', no such entry exists.]`
So, around 700 thousand data successfully created in the neo4j, the rest of them should work on 'worker01' node, but this node has the aforementioned error.
How this error occurred? 
Thanks.

**Comment by jexp** [comment]
Do you have a constraint on your Label(id) combination for your nodes?
Do you have a neo4j cluster or single instance?
What connection URL do you use.

Can you share code to reproduce?

**Comment by zhangwenxuan322** [comment]
1. I don't have a constraint on my nodes.
2. I'm using a neo4j cluster.
3. The URL that I config in spark:`spark.conf.set("spark.neo4j.url", "bolt://neo4j:pwd@ipaddress:port")`

The main process is simple. I have a dataTransfer function in Converter class which returns a nodes dataframe. Then I call the Neo4jDataFrame.createNodes() to create nodes into neo4j.
```
    val converter = new Converter(spark)
    val nodes = converter.dataTransfer()
    Neo4jDataFrame.createNodes(spark.sparkContext, nodes, ("shebei", nodes.columns.toSeq))
```
But I have two columns that may have `null` value in nodes dataframe. I don't know if this can cause the error.

**Comment by jexp** [comment]
Try bolt+routing 
And create the constraint

**Comment by jexp** [comment]
Oh sorry, as you don't pass merge=true you don't need the constraint.
But bolt+routing is required so that your write operations are properly routed in the cluster.

**Comment by zhangwenxuan322** [comment]
Thanks, I will try it later.

**Comment by jexp** [comment]
What kind of cluster are you running? You should run a causal cluster, not a HA cluster, otherwise the new drivers and bolt+routing won't work.

**Comment by zhangwenxuan322** [comment]
I'm running a causal cluster. After trying using a bolt+routing URL, I still got the same problem.

**Comment by zhangwenxuan322** [comment]
Actually, this error occurred sometimes among the whole writing process, but it didn't interrupt the program from running. So the consequence was that I lost some data. The main problem is this:`org.neo4j.driver.v1.exceptions.DatabaseException: LockResult[NOT_LOCKED, Unable to acquire shared lock: Cannot access 'RequestContext[machineId=2, lastAppliedTransaction=516, eventIdentifier=20, hashCode=1361372128, epoch=283068285884961, checksum=172509936730]', no such entry exists.]`

**Comment by zhangwenxuan322** [comment]
Now I'm thinking about saving data into hive and then using apoc write into neo4j.

**Comment by zhangwenxuan322** [comment]
Sorry, I created the whole nodes successfully at first. I made a mistake with the total number of data, so I didn't lose any data in neo4j. 
But I could only run my code successfully on neo4j-spark-connector:2.4.1-M1. However, on neo4j-spark-connector:2.4.5-M1, I was stuck on the repartition operation in Neo4jDataframe.execute().

**Comment by conker84** [comment]
> Sorry, I created the whole nodes successfully at first. I made a mistake with the total number of data, so I didn't lose any data in neo4j.
> But I could only run my code successfully on neo4j-spark-connector:2.4.1-M1. However, on neo4j-spark-connector:2.4.5-M1, I was stuck on the repartition operation in Neo4jDataframe.execute().

Can you try with 2.4.5-M1 and cache the input Dataframe before the `createNodes` operation?

**Comment by zhangwenxuan322** [comment]
@conker84 createNodes with 2.4.5-M1 works fine after caching the dataframe. But I still can't merge edges with `mergeEdgeList` function. Details in #135

**Comment by moxious** [comment]
Hi - I’m writing with good news.  Over the last few months, we’ve been investing a lot of effort in bringing the neo4j spark connector up to newer versions of spark, supporting the Datasource API, and enabling multi-language support for things like Python and R.

Today we’re putting out a pre-release of this new version of the spark connector.

You can download a build here:
https://github.com/neo4j-contrib/neo4j-spark-connector/releases/tag/4.0.0-pre1

And see the documentation here:
http://5.9.211.195:8000/neo4j-spark-docs/1.0.0/

We’re in the process of updating our guidance & documentation on spark.  We’re sorry, but we are unable fix issues & carry forward support of the earlier version of the connector, based on a different API - which went up to and included version 2.4.5-M2.  

Moving forward, we will be focusing our efforts on the new approach above.

As a result, we’re closing the remaining issues attached to the old version, but we would very much appreciate feedback & issues raised against the new 4.0 series of the connector, which you’ll find on the 4.0 branch of the repo.  The connector is going to stay open source, and our current issue list & github planning is in the open and will stay that way.  

We hope you find it a big improvement over the previous version,  and would welcome collaboration around this!
